File: LeetCode/LeetCode/container_with_most_water.go
package main

import "fmt"

func maxArea(height []int) int {
	i, j := 0 , len(height)-1
	max := 0

	for i < j {
		a , b := height[i], height[j]
		h := min(a, b)

		area := h * (j - i)
		if max < area {
			max = area
		}

		if a < b {
			i++
		}else {
			j--
		}
	}
	return max
}

func min(i, j int) int {
	if i <= j {
		return i
	}
	return j
}


func main() {
	fmt.Println(maxArea([]int{1,2,5,6}))
}

File: LeetCode/LeetCode/container_with_most_water.cc
#include<iostream>
#include<vector>
#include<algorithm>

using std::max;
using std::min;
using std::vector;

class Solution {
    public:
        int maxArea(vector<int>& height) {
            int i = 0, j = height.size() - 1, max_area = 0;
            while(i < j) {
                max_area = max(max_area, min(height[i], height[j]) * (j - i));
                if (height[i] > height[j]) {
                    --j;
                }else if(height[i] < height[j]) {
                    ++i;
                }else {
                    ++i, --j;
                }
            }
            return max_area;
        }
};

int main(void)
{
    vector<int> v{1,4,5,10};
    Solution s;
    std::cout<<s.maxArea(v)<<std::endl;
    return 0;
}

